File: cargo-cutoff-scraping/API/COSCO/cosco.js
import fs from 'fs';
import puppeteer from 'puppeteer-extra';
import stealth from 'puppeteer-extra-plugin-stealth';
import axios from "axios";

puppeteer.use(stealth());

async function cosco() {

    const browser = await puppeteer.launch({
        headless: false,
        defaultViewport: null,
    });
    
    const page = await browser.newPage();

    await page.setRequestInterception(true);

    let request_api = "https://elines.coscoshipping.com/ebschedule/public/purpoShipmentWs";

    const client = await page.createCDPSession();
    await client.send('Network.enable');

    client.on('Network.requestWillBeSent', async (params) => {
        if (params.request.url.includes('purpoShipmentWs')) {
            let cookies = await page.cookies();
            console.log("cookies ", cookies);
            fs.writeFileSync('cookie.json', JSON.stringify(cookies, null, 2));
            const cookieHeader = cookie();
            const requestHeaders = params.request.headers;
            requestHeaders['Cookie'] = cookieHeader;
            console.log('Request Headers:', requestHeaders);
            fs.writeFileSync('cosco_req_header.json', JSON.stringify(requestHeaders, null, 2));
            console.log('Request Payload:', params.request.postData);
            fs.writeFileSync('cosco_req_payload.json', JSON.stringify(params.request.postData, null, 2));

            function cookie(){
                const cookiesJson = JSON.parse(fs.readFileSync('cookie.json', 'utf-8'));
                const cookieHeader = cookiesJson.map(cookie => `${cookie.name}=${cookie.value}`).join('; ');
                return cookieHeader;
            }
        }
    });

    page.on('request', async (request) => {

        console.log(request.url());
        if (request.url().includes(request_api)) 
        {
            const requestHeaders = request.headers() | '';
            const requestPayload = request.postData() || '';

            console.log('Request Headers:', requestHeaders);
            fs.writeFileSync('cosco_req_header.json', JSON.stringify(requestHeaders, null, 2));

            console.log('Request Payload:', requestPayload);
            fs.writeFileSync('cosco_req_payload.json', JSON.stringify(requestPayload, null, 2));
        }

        request.continue();
    });

    try {
        await page.goto('https://elines.coscoshipping.com/ebusiness/sailingSchedule/searchByCity', { waitUntil: 'networkidle2' });
        console.log('Successfully navigated to the page!');
    } 
    catch (error) {
        console.error('Navigation error:', error);
    }


    const port_from = "Nhava sheva";
    const port_to = "Dublin";
    

    console.log("Entering port from...");

    await page.click('input.ivu-select-input[placeholder="Origin City (City,Province,Country/Region)"]');
    for (let i = 0; i < port_from.length; i++) {
        await page.keyboard.type(port_from[i], { delay: 500 });
    }

    // await page.waitForSelector('.ivu-select-dropdown', { visible: true, timeout: 3000 });
    await page.waitForSelector('body > div.app-wrapper.new-eb-layout > div > div.__panel > div > div.content > div.main-content > div:nth-child(1) > div > div > div > div > div > div > form > div:nth-child(1) > div.ivu-col.ivu-col-span-14 > div > div > div > div.ivu-select-dropdown > ul.ivu-select-dropdown-list > div', { visible: true, timeout: 3000 });
    await page.click('body > div.app-wrapper.new-eb-layout > div > div.__panel > div > div.content > div.main-content > div:nth-child(1) > div > div > div > div > div > div > form > div:nth-child(1) > div.ivu-col.ivu-col-span-14 > div > div > div > div.ivu-select-dropdown > ul.ivu-select-dropdown-list > div');
    await page.keyboard.press('ArrowDown');
    await page.keyboard.press('Enter');

    await page.evaluate(() => {
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve();
          }, 3000);
        });
    });

    console.log("Entering port to...");

    await page.click('input.ivu-select-input[placeholder="Destination City (City,Province,Country/Region)"]');
    for (let i = 0; i < port_to.length; i++) {
        await page.type('input.ivu-select-input[placeholder="Destination City (City,Province,Country/Region)"]', port_to[i], { delay: 500 });
    }

    await page.waitForSelector('body > div.app-wrapper.new-eb-layout > div > div.__panel > div > div.content > div.main-content > div:nth-child(1) > div > div > div > div > div > div > form > div:nth-child(2) > div.ivu-col.ivu-col-span-14 > div > div > div > div.ivu-select-dropdown > ul.ivu-select-dropdown-list > div', { visible: true , timeout: 3000});
    await page.click('body > div.app-wrapper.new-eb-layout > div > div.__panel > div > div.content > div.main-content > div:nth-child(1) > div > div > div > div > div > div > form > div:nth-child(2) > div.ivu-col.ivu-col-span-14 > div > div > div > div.ivu-select-dropdown > ul.ivu-select-dropdown-list > div');
    await page.keyboard.press('ArrowDown');
    await page.keyboard.press('Enter');

    await page.waitForSelector('.btnSearch.ivu-btn', { timeout: 5000 });

    await page.click('.btnSearch.ivu-btn');  

    await page.evaluate(() => {
        return new Promise((resolve) => {
          setTimeout(() => {
            resolve();
          }, 30000);
        });
    });


    const url = 'https://elines.coscoshipping.com/ebschedule/public/purpoShipmentWs';

    const headers = JSON.parse(fs.readFileSync('cosco_req_header.json', 'utf-8'));

    const body = JSON.parse(fs.readFileSync('cosco_req_payload.json', 'utf-8'));

    try {
        const response = await axios.post(url, body, {
            headers: headers,
        });
        console.log('Response data:', response.data);
        fs.writeFileSync('cosco_rawdata.json', JSON.stringify(response.data, null, 2));

    } catch (error) {
        console.error('Error sending request:', error.message);
    }

    await browser.close();
}

cosco();

File: scraping-cargocutoff/cosco.py
from selenium_driverless import webdriver
from selenium_driverless.types.by import By
import asyncio
import json
import requests
import  time

async def cosco():
    options = webdriver.ChromeOptions()

    driver = webdriver.Chrome(options=options)

    targetUrl = "https://elines.coscoshipping.com/ebschedule/public/purpoShipmentWs"

    request_data = {
        "url": '',
        "all_request_cookies": None,
        "request_headers": None,
        "request_payload": None
    }

    timeout_seconds = 300
    async with driver as browser:
        try:

            await browser.get('https://elines.coscoshipping.com/ebusiness/sailingSchedule/searchByCity', wait_load=True, timeout=timeout_seconds)
            await browser.sleep(2)

            print("Successfully navigated to the page!")
            # cookie_button = await browser.find_element(By.ID, 'onetrust-accept-btn-handler', timeout=10)
            # if cookie_button:
            #     print("Cookie consent button found.")
            #     await cookie_button.click(move_to=True)
            # else:
            #     print("Cookie consent button not found.")
            #
            # # await driver.wait_for_cdp("Page.domContentEventFired", timeout=15)
            #
            await browser.execute_cdp_cmd('Network.enable', {})

            async def capture_request(request):
                data = request
                print(data['request']['url'])
                if targetUrl in data['request']['url']:
                    request_data['url'] = data['request']['url']
                    request_data['request_headers'] = data['request']['headers']
                    print(f"Request matched the target URL: {data['request']['url']}")
                    print(f"Request: {json.dumps(data, indent=2)}")
                    print(f"Request Headers: {json.dumps(data['request']['headers'], indent=2)}")
                    request_data['request_payload'] = data['request']['postData']
                    # if 'postData' in request:
                    #     request_data['request_payload'] = data['request']['postData']
                    #     print(f"Payload: {data['request']['postDataEntries']}")

                    cookies = await browser.get_cookies()
                    cookies_str = "; ".join([f"{cookie['name']}={cookie['value']}" for cookie in cookies])
                    request_data['all_request_cookies'] = cookies_str
                    print(f"Cookies: {json.dumps(cookies, indent=2)}")

            await browser.add_cdp_listener('Network.requestWillBeSent', capture_request)

            port_from = "Nhava Sheva"
            pol = await browser.find_element(By.CSS_SELECTOR, 'input.ivu-select-input[placeholder="Origin City (City,Province,Country/Region)"]', timeout=3)
            await pol.focus()
            await pol.send_keys(port_from)
            await browser.sleep(5)
            dropdown_pol = await browser.find_element(By.CSS_SELECTOR, 'body > div.app-wrapper.new-eb-layout > div > div.__panel > div > div.content > div.main-content > div:nth-child(1) > div > div > div > div > div > div > form > div:nth-child(1) > div.ivu-col.ivu-col-span-14 > div > div > div > div.ivu-select-dropdown > ul.ivu-select-dropdown-list > div', timeout=5)
            await dropdown_pol.click(move_to=True)
            await asyncio.sleep(5)

            port_to = "Singapore"
            pod = await browser.find_element(By.CSS_SELECTOR, 'input.ivu-select-input[placeholder="Destination City (City,Province,Country/Region)"]', timeout=3)
            await pod.focus()
            await pod.send_keys(port_to)
            await browser.sleep(5)
            dropdown_pod = await browser.find_element(By.CSS_SELECTOR, 'body > div.app-wrapper.new-eb-layout > div > div.__panel > div > div.content > div.main-content > div:nth-child(1) > div > div > div > div > div > div > form > div:nth-child(2) > div.ivu-col.ivu-col-span-14 > div > div > div > div.ivu-select-dropdown > ul.ivu-select-dropdown-list > div', timeout=5)
            await dropdown_pod.click(move_to=True)
            await asyncio.sleep(5)
            # #
            # # search_card = await browser.find_element(By.CSS_SELECTOR, '.cards-search-button', timeout=5)
            # # print(search_card)
            search_scedule = await browser.find_element(By.CSS_SELECTOR, '.btnSearch.ivu-btn', timeout=5)
            if search_scedule:
                print("Search button found.")
                await search_scedule.click(move_to=True)
                await asyncio.sleep(3)
            else:
                print("Search button not found.")

            await asyncio.sleep(3)

            print("Captured Cookies:", request_data['all_request_cookies'])

            print("Captured Headers:", json.dumps(request_data['request_headers'], indent=2))
            headers_with_cookie = request_data['request_headers'].copy()
            headers_with_cookie['Cookie'] = request_data['all_request_cookies']  # Add the cookies string to headers
            print(f"Updated Request Headers with Cookie: {json.dumps(headers_with_cookie, indent=2)}")
            print("Captured Payload:", json.dumps(request_data['request_payload'], indent=2))
            print("request url:", request_data['url'])

            try:
                print(request_data['url'])
                print(request_data['request_payload'])
                print(headers_with_cookie)
                payload = json.loads(request_data['request_payload'])
                print("after updated:",payload)
                response = requests.request("POST", request_data['url'], json=payload, headers=headers_with_cookie)
                # response = requests.post(request_data['url'], json=request_data['request_payload'], headers=headers_with_cookie)
                print(f"Status Code: {response.status_code}")
                await asyncio.sleep(3)
                # response.raise_for_status()
                print('Response text:', response.text)
                print('Response data:', response.json())

                with open('cosco_rawdata.json', 'w') as json_file:
                    json.dump(response.json(), json_file, indent=2)

            except requests.exceptions.RequestException as e:
                print(f"Error: {e}")


        except Exception as e:
            print(f"Navigation error: {e}")
            browser.quit()
            return

        await browser.quit()


if __name__ == "__main__":
    asyncio.run(cosco())